File: src/main/scala/uk/co/ridentbyte/model/Curl.scala
package uk.co.ridentbyte.model

import uk.co.ridentbyte.Cardinal

case class Curl(uri: String, verb: String, body: Option[String], headers: List[String], envVars: List[EnvironmentVariable]) {

  def toCommand: String = {
    val sb = new StringBuilder
    sb.append("curl -i \\\n")

    headers.foreach { header =>
      sb.append(s"""-H '${RequestString(header, envVars, Cardinal.vocabulary).process}' \\\n""")
    }

    body.foreach { b =>
      sb.append(s"""-d '${RequestString(b, envVars, Cardinal.vocabulary).process}' \\\n""")
    }

    sb.append(s"""-X $verb ${RequestString(uri, envVars, Cardinal.vocabulary).process}""")

    sb.toString()
  }

}

File: src/main/java/uk/co/ridentbyte/model/Curl.java
package uk.co.ridentbyte.model;

import java.util.List;

public class Curl {

    private final boolean followRedirects;
    private String uri, verb, body;
    private List<String> headers;
    private List<EnvironmentVariable> envVars;
    private Vocabulary vocabulary;

    public Curl(String uri,
                String verb,
                boolean followRedirects,
                String body,
                List<String> headers,
                List<EnvironmentVariable> envVars,
                Vocabulary vocabulary) {
        this.uri = uri;
        this.verb = verb;
        this.followRedirects = followRedirects;
        this.body = body;
        this.headers = headers;
        this.envVars = envVars;
        this.vocabulary = vocabulary;
    }

    public String toCommand() {
        StringBuilder builder = new StringBuilder("curl -i \\\n");
        if (followRedirects) {
            builder.append("-L ");
        }

        for (String header : this.headers) {
            String processedString = new RequestString(header, this.envVars, vocabulary).process();
            builder.append("-H '");
            builder.append(processedString);
            builder.append("' \\\n");
        }

        if (this.body != null && !this.body.isEmpty()) {
            String processedString = new RequestString(this.body, this.envVars, vocabulary).process();
            builder.append("-d '");
            builder.append(processedString);
            builder.append("' \\\n");
        }

        String processedString = new RequestString(this.uri, this.envVars, vocabulary).process();
        builder.append("-X ");
        builder.append(this.verb);
        builder.append(" ");
        builder.append(processedString);
        return builder.toString();
    }

}
